src/binary/raw_binary_reader.rs

```diff
@@ -259,7 +261,7 @@ macro_rules! read_safety_checks {
         if $raw_binary_reader.cursor.value.ion_type != $ion_type
             || $raw_binary_reader.cursor.value.is_null
         {
-            return Ok(None);
+            return Err($raw_binary_reader.expected(format!("{} value", $ion_type)));
```

Comment:
Calling `read_string()` on an `int` (or other type-mismatched operations) now returns `IonError::IllegalOperation` instead of `Ok(None)`.